### DISCUSSION TITLE
Localization

### DISCUSSION THREAD

**Discussion by dnno** [CappielloAntonio/tempo#27] [Ideas]
How do you feel about localizations for Tempo? It's fairly easy to do, provided one knows the corresponding language. I have created a localized german version of Tempo today and I could open a PR for that, if you are interested. I'm sure there are others who could chip in with more languages, making Tempo accessible to a greater userbase.

**Comment by CappielloAntonio** [comment]
Hey thanks a lot for the offer! 
Yes, localization is in my top five priorities right now. I would also like to take the opportunity to learn about the use of some online tools to automate or at least facilitate translation into new languages.
Do you recommend any tools in particular?

First of all though I'd have to port a few hardcoded strings from the code into the string.xml file.

**Reply by dnno** [reply]
I have been working with weblate when working on a different app. It makes the process accessible to non-technical folks. But you'd need to run weblate somewhere, so there's some overhead here. In the end it just updates the resource files, so you could always add weblate to the team later, even if you start localizing manually.

Self-hosting is free, and the setup is fairly simple I think: https://docs.weblate.org/en/latest/formats/android.html

Please tell me if you think I can help with something!

**Reply by dnno** [reply]
A small update: I have set up a private weblate server for my fork of your repository [using docker compose](https://docs.weblate.org/en/latest/admin/install/docker.html). While I have not fully configured it, it seems to correctly pick up on the two lanugages present in my fork (English, German). There's probably some details left to setup (like webhooks and creating PRs against Github to integrate changes to the localizations, but I think the effort is managable and well documented.

One thing that I discovered: Weblate currently doesn't work with string-arrays and you use those in some cases. There's a workaround, however, prominently featured on the Android documentaiton page: https://docs.weblate.org/en/latest/formats/android.html.

I know this is not your top-priority, but let me know when you decide to work on this and I can help.

**Comment by dnno** [comment]
@CappielloAntonio, how do you feel about a "language switcher" in Tempo? Normally, Android automatically chooses the appropriate locale for a user, falling back to the default locale. I think it might be handy to have a preference, that enables a user to switch between supported languages.
I have played around a bit and my approach would look something like this (screenshots from settings view, as you surely recognize). If you find it useful, I could polish my solution a bit and open a PR.

![Screenshot_2023-08-16-21-16-01-93_8fff24685b220ab505c5c7f79d4b9bbf](https://github.com/CappielloAntonio/tempo/assets/2003552/0d9bd651-80ca-4889-99b6-4c2d88cde825)
![Screenshot_2023-08-16-21-16-11-15_8fff24685b220ab505c5c7f79d4b9bbf](https://github.com/CappielloAntonio/tempo/assets/2003552/2179decc-bdf7-4e72-a2ed-22129779e080)
![Screenshot_2023-08-16-21-16-21-93_8fff24685b220ab505c5c7f79d4b9bbf](https://github.com/CappielloAntonio/tempo/assets/2003552/ee97b047-a43a-4148-9a22-d6f93dc47af6)
![Screenshot_2023-08-16-21-16-36-99_8fff24685b220ab505c5c7f79d4b9bbf](https://github.com/CappielloAntonio/tempo/assets/2003552/6a678d4d-9515-4c1b-9900-48f276353266)
![Screenshot_2023-08-16-21-16-49-09_8fff24685b220ab505c5c7f79d4b9bbf](https://github.com/CappielloAntonio/tempo/assets/2003552/62d9fc96-b617-4ba5-9e2c-275b62dd8eaf)

**Comment by CappielloAntonio** [comment]
Hi @dnno, generally speaking I've always thought that the system language should "command" the language used in the app. Along these lines I would like to give the user not too many options to play with, so a language switcher was not in my priorities.

However, one of the changes introduced only recently by the Android [new API](https://developer.android.com/guide/topics/resources/app-languages#androidx-impl) that I would have liked to explore is the [preference of per-app language](https://developer.android.com/guide/topics/resources/app-languages).

I have a vague idea of how to implement this function: it shouldn't be too complicated, but as I said before it wasn't one of my properties. May I ask how did you implement it? How did you retrieve the list of available languages? Does the application require an app restart or is it applied immediately?

Anyways, nice job! I'm glad to know that someone other than me is even getting their hands on the code after nearly three years of solitary development.

**Reply by dnno** [reply]
I have opened a [draft PR](https://github.com/CappielloAntonio/tempo/pull/38) for your reference, there are probably some issues with it but it works on first sight.

The languages are retrieved from the application resources. The app needs a restart with my implementation, i'm not 100% sure if that's absolutely necessary, but I could do further research on that...

You did an amazing job in those three years, let me tell you that!

**Reply by CappielloAntonio** [reply]
Maybe I can help you with your research in this dead time. 

I'm in a stalemate, I'm waiting for the test results on this round of implementations and the next activity would be the Android Auto implementation (it seems to be the latest issue at the moment) but it's a bit scary.

This language switch can be an entertaining distraction.

**Comment by CappielloAntonio** [comment]
Hey @dnno, I'm about to push my work to the language picker from app settings and system settings. Maybe I didn't find the correct documentation, maybe it's the official documentation lacking clarity, but I jumped a bit to get the correct list of supported languages.

In your implementation, the list of languages shown to me were those that the system supports (trivially, the list of languages that the keyboard supports in multilanguage). In my case there was Italian and American English, but not German, in your case English and German (which coincidentally are the two really supported languages).

To get the real list of languages (which corresponds to the list in the app's system settings) I had to work a bit.

Soon I'll put the changes online, tell me what you think!

**Comment by CappielloAntonio** [comment]
Done and made sure it works up to Android API version 26

**Reply by dnno** [reply]
Very cool, thanks for the effort!

**Comment by rkowalsk** [comment]
Hey, I'd like to add french localization but I've never done android development before. Apart from creating the `app/src/main/res/values-fr/` folder with `arrays.xml` and `strings.xml` with the right translations and adding `<locale android:name="fr-FR"/>` to `app/src/main/res/xml/locale_config.xml`, what needs to be done ?
Of course if you have a working Weblate instance I'd happily use it !

**Reply by dnno** [reply]
That's all you need to do. I can do a quick check if the app seems correctly translated if you push your changes to your GitHub.
To my knowledge, Weblate isn't available (yet?)...
